In [18]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, r2_score

In [35]:
data_train = pd.read_csv('cs-training.csv')
data_train.head(100)

Unnamed: 0  SeriousDlqin2yrs  RevolvingUtilizationOfUnsecuredLines  age  \
0            1                 1                              0.766127   45   
1            2                 0                              0.957151   40   
2            3                 0                              0.658180   38   
3            4                 0                              0.233810   30   
4            5                 0                              0.907239   49   
..         ...               ...                                   ...  ...   
95          96                 0                              0.245353   37   
96          97                 0                              0.542243   48   
97          98                 0                              0.010531   57   
98          99                 0                              0.363200   32   
99         100                 0                              0.032618   75   

    NumberOfTime30-59DaysPastDueNotWorse  DebtRatio  MonthlyIncome  \
0                                      2   0.802982         9120.0   
1                                      0   0.121876         2600.0   
2                                      1   0.085113         3042.0   
3                                      0   0.036050         3300.0   
4                                      1   0.024926        63588.0   
..                                   ...        ...            ...   
95                                     0   0.288417         6500.0   
96                                     2  10.000000            NaN   
97                                     0   0.280665         5714.0   
98                                     0   0.480524         2900.0   
99                                     0   0.006799         7500.0   

    NumberOfOpenCreditLinesAndLoans  NumberOfTimes90DaysLate  \
0                                13                        0   
1                                 4                        0   
2                                 2                        1   
3                                 5                        0   
4                                 7                        0   
..                              ...                      ...   
95                               11                        1   
96                                2                        0   
97                                6                        0   
98                                4                        0   
99                                4                        0   

    NumberRealEstateLoansOrLines  NumberOfTime60-89DaysPastDueNotWorse  \
0                              6                                     0   
1                              0                                     0   
2                              0                                     0   
3                              0                                     0   
4                              1                                     0   
..                           ...                                   ...   
95                             1                                     1   
96                             0                                     0   
97                             1                                     0   
98                             1                                     0   
99                             0                                     0   

    NumberOfDependents  
0                  2.0  
1                  1.0  
2                  0.0  
3                  0.0  
4                  0.0  
..                 ...  
95                 0.0  
96                 NaN  
97                 0.0  
98                 0.0  
99                 0.0  

[100 rows x 12 columns]

In [27]:
data_test = pd.read_csv('cs-test.csv')
data_test.head(10)

Unnamed: 0  SeriousDlqin2yrs  RevolvingUtilizationOfUnsecuredLines  age  \
0           1               NaN                              0.885519   43   
1           2               NaN                              0.463295   57   
2           3               NaN                              0.043275   59   
3           4               NaN                              0.280308   38   
4           5               NaN                              1.000000   27   
5           6               NaN                              0.509791   63   
6           7               NaN                              0.587778   50   
7           8               NaN                              0.046149   79   
8           9               NaN                              0.013527   68   
9          10               NaN                              1.000000   23   

   NumberOfTime30-59DaysPastDueNotWorse    DebtRatio  MonthlyIncome  \
0                                     0     0.177513         5700.0   
1                                     0     0.527237         9141.0   
2                                     0     0.687648         5083.0   
3                                     1     0.925961         3200.0   
4                                     0     0.019917         3865.0   
5                                     0     0.342429         4140.0   
6                                     0  1048.000000            0.0   
7                                     1     0.369170         3301.0   
8                                     0  2024.000000            NaN   
9                                    98     0.000000            0.0   

   NumberOfOpenCreditLinesAndLoans  NumberOfTimes90DaysLate  \
0                                4                        0   
1                               15                        0   
2                               12                        0   
3                                7                        0   
4                                4                        0   
5                                4                        0   
6                                5                        0   
7                                8                        0   
8                                4                        0   
9                                0                       98   

   NumberRealEstateLoansOrLines  NumberOfTime60-89DaysPastDueNotWorse  \
0                             0                                     0   
1                             4                                     0   
2                             1                                     0   
3                             2                                     0   
4                             0                                     0   
5                             0                                     0   
6                             0                                     0   
7                             1                                     0   
8                             1                                     0   
9                             0                                    98   

   NumberOfDependents  
0                 0.0  
1                 2.0  
2                 2.0  
3                 0.0  
4                 1.0  
5                 1.0  
6                 3.0  
7                 1.0  
8                 0.0  
9                 0.0

In [29]:
data_probability = pd.read_csv('sampleEntry.csv')
data_probability.head(10)

Id  Probability
0   1     0.080807
1   2     0.040719
2   3     0.011968
3   4     0.067640
4   5     0.108264
5   6     0.027520
6   7     0.047294
7   8     0.041027
8   9     0.004223
9  10     0.377129

# Обработка данных

In [30]:
# Пропуски в данных тестовых
data_train.isnull().sum()

Unnamed: 0                                  0
SeriousDlqin2yrs                            0
RevolvingUtilizationOfUnsecuredLines        0
age                                         0
NumberOfTime30-59DaysPastDueNotWorse        0
DebtRatio                                   0
MonthlyIncome                           29731
NumberOfOpenCreditLinesAndLoans             0
NumberOfTimes90DaysLate                     0
NumberRealEstateLoansOrLines                0
NumberOfTime60-89DaysPastDueNotWorse        0
NumberOfDependents                       3924
dtype: int64

In [31]:
# Пропуски в данных тренеровочных
data_test.isnull().sum()

Unnamed: 0                                   0
SeriousDlqin2yrs                        101503
RevolvingUtilizationOfUnsecuredLines         0
age                                          0
NumberOfTime30-59DaysPastDueNotWorse         0
DebtRatio                                    0
MonthlyIncome                            20103
NumberOfOpenCreditLinesAndLoans              0
NumberOfTimes90DaysLate                      0
NumberRealEstateLoansOrLines                 0
NumberOfTime60-89DaysPastDueNotWorse         0
NumberOfDependents                        2626
dtype: int64

In [32]:
# Пропуски в данных вероятности
data_probability.isnull().sum()

Id             0
Probability    0
dtype: int64

Для тренеровочных данных

In [40]:
data_train.groupby(data_train['MonthlyIncome'].isnull()).mean()

Unnamed: 0  SeriousDlqin2yrs  \
MonthlyIncome                                   
False          75026.012514          0.069486   
True           74897.295786          0.056137   

               RevolvingUtilizationOfUnsecuredLines        age  \
MonthlyIncome                                                    
False                                      5.899873  51.289792   
True                                       6.649421  56.362349   

               NumberOfTime30-59DaysPastDueNotWorse    DebtRatio  \
MonthlyIncome                                                      
False                                      0.381769    26.598777   
True                                       0.579866  1673.396556   

               MonthlyIncome  NumberOfOpenCreditLinesAndLoans  \
MonthlyIncome                                                   
False            6670.221237                         8.758475   
True                     NaN                         7.216071   

               NumberOfTimes90DaysLate  NumberRealEstateLoansOrLines  \
MonthlyIncome                                                          
False                         0.211925                      1.054519   
True                          0.484612                      0.871481   

               NumberOfTime60-89DaysPastDueNotWorse  NumberOfDependents  
MonthlyIncome                                                            
False                                      0.187829            0.851832  
True                                       0.452995            0.316310

Видно, что относительно тех лиц, месячная зарплата которых указана, в среднем: (Y/N); возвраст 51 год; округляя 6% людей с постоянным использованием карт; до 1 человека, кто может пропустить дедлайн от 30 до 59 дней; коэффициент задолженности ~27% ; их месячная зарпрала в средне 6670$; кол-во открытых кредитов ~9; кол-во опозданий за оплату на 90 дней и более 0,2; кол-во кредитов на недвижимость ~1; кол-во иждевенцев в семье ~1; кол-во дней просрочки от 60 до 89 дней = 0,18.

Относительно тех, чья месячная зарплата не указана, видно, что в среднес: (Y/N); возвраст 56 лет; округляя 7% людей с постоянным использованием кредитных карт; до 1 человека кто может пропустить просрочку от 30 до 59 дней; коэффициент задолженности 1673%(!!!); кол-во открытых кредитов ~8; кол-во людей с просрочко на 90 дней 0,5; кол-во кредитов на недвижимость ~ 1%; 

In [42]:
data_train.groupby(data_train['NumberOfDependents'].isnull()).mean()

Unnamed: 0  SeriousDlqin2yrs  \
NumberOfDependents                                   
False               75013.668597          0.067410   
True                74510.281855          0.045617   

                    RevolvingUtilizationOfUnsecuredLines        age  \
NumberOfDependents                                                    
False                                           5.922272  52.099277   
True                                           10.745132  59.588940   

                    NumberOfTime30-59DaysPastDueNotWorse    DebtRatio  \
NumberOfDependents                                                      
False                                           0.407945   333.373603   
True                                            0.908257  1083.812181   

                    MonthlyIncome  NumberOfOpenCreditLinesAndLoans  \
NumberOfDependents                                                   
False                 6670.221237                         8.529279   
True                          NaN                         5.604230   

                    NumberOfTimes90DaysLate  NumberRealEstateLoansOrLines  \
NumberOfDependents                                                          
False                              0.250698                      1.029717   
True                               0.834608                      0.590979   

                    NumberOfTime60-89DaysPastDueNotWorse  NumberOfDependents  
NumberOfDependents                                                            
False                                           0.225027            0.757222  
True                                            0.812181                 NaN

Для тестовых данных

In [46]:
data_test.groupby(data_test['SeriousDlqin2yrs'].isnull()).mean()

Unnamed: 0  SeriousDlqin2yrs  \
SeriousDlqin2yrs                                 
True                 50752.0               NaN   

                  RevolvingUtilizationOfUnsecuredLines        age  \
SeriousDlqin2yrs                                                    
True                                              5.31  52.405436   

                  NumberOfTime30-59DaysPastDueNotWorse  DebtRatio  \
SeriousDlqin2yrs                                                    
True                                           0.45377  344.47502   

                  MonthlyIncome  NumberOfOpenCreditLinesAndLoans  \
SeriousDlqin2yrs                                                   
True                 6855.03559                         8.453514   

                  NumberOfTimes90DaysLate  NumberRealEstateLoansOrLines  \
SeriousDlqin2yrs                                                          
True                             0.296691                      1.013074   

                  NumberOfTime60-89DaysPastDueNotWorse  NumberOfDependents  
SeriousDlqin2yrs                                                            
True                                          0.270317            0.769046

In [47]:
data_test.groupby(data_test['MonthlyIncome'].isnull()).mean()

Unnamed: 0  SeriousDlqin2yrs  \
MonthlyIncome                                   
False          50735.339324               NaN   
True           50819.461523               NaN   

               RevolvingUtilizationOfUnsecuredLines        age  \
MonthlyIncome                                                    
False                                      5.123221  51.347260   
True                                       6.066298  56.690146   

               NumberOfTime30-59DaysPastDueNotWorse    DebtRatio  \
MonthlyIncome                                                      
False                                      0.400319    28.828501   
True                                       0.670198  1622.574143   

               MonthlyIncome  NumberOfOpenCreditLinesAndLoans  \
MonthlyIncome                                                   
False             6855.03559                         8.749349   
True                     NaN                         7.255633   

               NumberOfTimes90DaysLate  NumberRealEstateLoansOrLines  \
MonthlyIncome                                                          
False                         0.230885                      1.050848   
True                          0.563150                      0.860120   

               NumberOfTime60-89DaysPastDueNotWorse  NumberOfDependents  
MonthlyIncome                                                            
False                                      0.205651            0.864988  
True                                       0.532159            0.322195

In [48]:
data_test.groupby(data_test['NumberOfDependents'].isnull()).mean()

Unnamed: 0  SeriousDlqin2yrs  \
NumberOfDependents                                   
False               50744.615421               NaN   
True                51030.052171               NaN   

                    RevolvingUtilizationOfUnsecuredLines        age  \
NumberOfDependents                                                    
False                                           5.346712  52.214721   
True                                            3.927689  59.586443   

                    NumberOfTime30-59DaysPastDueNotWorse   DebtRatio  \
NumberOfDependents                                                     
False                                           0.439364  328.824671   
True                                            0.996192  933.758949   

                    MonthlyIncome  NumberOfOpenCreditLinesAndLoans  \
NumberOfDependents                                                   
False                  6855.03559                         8.528677   
True                          NaN                         5.623382   

                    NumberOfTimes90DaysLate  NumberRealEstateLoansOrLines  \
NumberOfDependents                                                          
False                              0.279964                      1.026063   
True                               0.926504                      0.523991   

                    NumberOfTime60-89DaysPastDueNotWorse  NumberOfDependents  
NumberOfDependents                                                            
False                                           0.253881            0.769046  
True                                            0.889185                 NaN

Выводы: месячную зарплату проставлять 0, так как их коэффициент задолженности составляет 1673%. Колисетво иждевенцев в среднем одно и тоже значение, значит заполнить пропуски средним значение. А предсказывать нужно признак SeriousDlqin2yrs

In [54]:
# Заполним 0 месячную зарплату в тренеровочной выборке 
data_train['MonthlyIncome'] = data_train['MonthlyIncome'].fillna(0)
print(data_train.head(10))

   Unnamed: 0  SeriousDlqin2yrs  RevolvingUtilizationOfUnsecuredLines  age  \
0           1                 1                              0.766127   45   
1           2                 0                              0.957151   40   
2           3                 0                              0.658180   38   
3           4                 0                              0.233810   30   
4           5                 0                              0.907239   49   
5           6                 0                              0.213179   74   
6           7                 0                              0.305682   57   
7           8                 0                              0.754464   39   
8           9                 0                              0.116951   27   
9          10                 0                              0.189169   57   

   NumberOfTime30-59DaysPastDueNotWorse    DebtRatio  MonthlyIncome  \
0                                     2     0.802982         9120.0   

In [55]:
# Заполним средним количество иждевенцев в семье
data_train['NumberOfDependents'].fillna(data_train['NumberOfDependents'].mean(), inplace = True)

In [56]:
data_train.head(20)

Unnamed: 0  SeriousDlqin2yrs  RevolvingUtilizationOfUnsecuredLines  age  \
0            1                 1                              0.766127   45   
1            2                 0                              0.957151   40   
2            3                 0                              0.658180   38   
3            4                 0                              0.233810   30   
4            5                 0                              0.907239   49   
5            6                 0                              0.213179   74   
6            7                 0                              0.305682   57   
7            8                 0                              0.754464   39   
8            9                 0                              0.116951   27   
9           10                 0                              0.189169   57   
10          11                 0                              0.644226   30   
11          12                 0                              0.018798   51   
12          13                 0                              0.010352   46   
13          14                 1                              0.964673   40   
14          15                 0                              0.019657   76   
15          16                 0                              0.548458   64   
16          17                 0                              0.061086   78   
17          18                 0                              0.166284   53   
18          19                 0                              0.221813   43   
19          20                 0                              0.602794   25   

    NumberOfTime30-59DaysPastDueNotWorse    DebtRatio  MonthlyIncome  \
0                                      2     0.802982         9120.0   
1                                      0     0.121876         2600.0   
2                                      1     0.085113         3042.0   
3                                      0     0.036050         3300.0   
4                                      1     0.024926        63588.0   
5                                      0     0.375607         3500.0   
6                                      0  5710.000000            0.0   
7                                      0     0.209940         3500.0   
8                                      0    46.000000            0.0   
9                                      0     0.606291        23684.0   
10                                     0     0.309476         2500.0   
11                                     0     0.531529         6501.0   
12                                     0     0.298354        12454.0   
13                                     3     0.382965        13700.0   
14                                     0   477.000000            0.0   
15                                     0     0.209892        11362.0   
16                                     0  2058.000000            0.0   
17                                     0     0.188274         8800.0   
18                                     0     0.527888         3280.0   
19                                     0     0.065868          333.0   

    NumberOfOpenCreditLinesAndLoans  NumberOfTimes90DaysLate  \
0                                13                        0   
1                                 4                        0   
2                                 2                        1   
3                                 5                        0   
4                                 7                        0   
5                                 3                        0   
6                                 8                        0   
7                                 8                        0   
8                                 2                        0   
9                                 9                        0   
10                                5                        0   
11                                7                        0   

In [57]:
data_train.isnull().sum()

Unnamed: 0                              0
SeriousDlqin2yrs                        0
RevolvingUtilizationOfUnsecuredLines    0
age                                     0
NumberOfTime30-59DaysPastDueNotWorse    0
DebtRatio                               0
MonthlyIncome                           0
NumberOfOpenCreditLinesAndLoans         0
NumberOfTimes90DaysLate                 0
NumberRealEstateLoansOrLines            0
NumberOfTime60-89DaysPastDueNotWorse    0
NumberOfDependents                      0
dtype: int64

In [ ]:
# Отделяем признаки и целевую переменную
# В признаки добавим то, что инетер
X = data[['NumberOfTimes90DaysLate','DebtRatio','']]
y = data['SeriousDlqin2yrs']
